segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameBuilder.java

```diff
@@ -9,36 +9,43 @@
  */
 package io.pravega.segmentstore.server.logs;
 
+import com.google.common.base.Preconditions;
 import io.pravega.common.ExceptionHelpers;
 import io.pravega.common.Exceptions;
-import io.pravega.common.function.CallbackHelpers;
-import io.pravega.common.function.ConsumerWithException;
+import io.pravega.common.ObjectClosedException;
+import io.pravega.common.util.SortedIndex;
 import io.pravega.segmentstore.server.LogItem;
 import io.pravega.segmentstore.storage.DurableDataLog;
 import io.pravega.segmentstore.storage.LogAddress;
-import com.google.common.base.Preconditions;
-import lombok.extern.slf4j.Slf4j;
-
 import java.io.IOException;
 import java.time.Duration;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
 import java.util.function.Consumer;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
 
 /**
  * Builds DataFrames from LogItems. Splits the serialization of LogItems across multiple Data Frames, if necessary,
  * and publishes the finished Data Frames to the given DataFrameLog.
  */
 @Slf4j
+@NotThreadSafe
 class DataFrameBuilder<T extends LogItem> implements AutoCloseable {
     //region Members
 
-    private static final Duration DATA_FRAME_WRITE_TIMEOUT = Duration.ofSeconds(30); // TODO: actual timeout.
     private final DataFrameOutputStream outputStream;
     private final DurableDataLog targetLog;
-    private final ConsumerWithException<DataFrameCommitArgs, Exception> dataFrameCommitSuccessCallback;
-    private final Consumer<Throwable> dataFrameCommitFailureCallback;
-    private boolean closed;
+    private final Args args;
+    private final AtomicBoolean closed;
     private long lastSerializedSequenceNumber;
     private long lastStartedSequenceNumber;
+    private final AtomicReference<Throwable> failureCause;
```

Comment:
I don't want to add locks to append and flush since that has the potential to deadlock in case of failures (catch block happen with the lock acquired, and if the callback itself needs to acquire a lock externally (and mostly it does), and it's on a different thread, they will deadlock. My initial implementation was to have DataFrameBuilder 100% thread safe, but I stumbled across these problems so much that I decided to simplify it.